python/cudf/dataframe/index.py

```diff
@@ -393,4 +382,86 @@ def get_dt_field(self, field):
                                      mask=out_column.mask,
                                      null_count=out_column.null_count,
                                      dtype=out_column.dtype)
-        return GenericIndex(out_column)
+        return as_index(out_column)
+
+
+class CategoricalIndex(GenericIndex):
+    """An categorical of orderable values that represent the indices of another
+    Column
+
+    Attributes
+    ---
+    _values: A CategoricalColumn object
+    name: A string
+    """
+    def __init__(self, values, name=None):
+        if isinstance(values, pd.Series) and \
+           pd.api.types.is_categorical_dtype(values.dtype):
+            values = CategoricalColumn(
+                data=Buffer(values.cat.codes.values),
+                categories=values.cat.categories.tolist(),
+                ordered=values.cat.ordered
+            )
+        elif isinstance(values, (pd.Categorical, pd.CategoricalIndex)):
+            values = CategoricalColumn(
+                data=Buffer(values.codes),
+                categories=values.categories.tolist(),
+                ordered=values.ordered
+            )
+
+        self._values = values
+        self.name = name
+        self.names = [name]
+
+    @property
+    def codes(self):
+        return self._values.codes
+
+    @property
+    def categories(self):
+        return self._values.categories
+
+
+def as_index(arbitrary, name=None):
+    """Create an Index from an arbitrary object
+
+    Currently supported inputs are:
+
+    * ``Column``
+    * ``Buffer``
+    * ``Series``
+    * ``Index``
+    * numba device array
+    * numpy array
+    * pyarrow array
+    * pandas.Categorical
+
+    Returns
+    -------
+    result : subclass of Index
+        - CategoricalIndex for Categorical input.
+        - DatetimeIndex for Datetime input.
+        - GenericIndex for all other inputs.
+    """
+    # This function should probably be moved to Index.__new__
```

Comment:
Any particular reason why its not in `Index.__new__` yet?

Reply:
I kinda followed the pattern of `columnops.as_column` for now, no particularly strong reason, but I'd like to hold off on further refactoring until 0.6+.

Reply:
fine by me, I was just curious

Reply:
👍